### PR TITLE
Custom headers for jsonRpcProvider

### PR DESCRIPTION
**Pull request by IvanLudvig**
adds functionality to pass headers as arguments to jsonRpcProvider. this is useful as some rpcs require passing the auth token in a header

### CODE REVIEW

File: packages/chains/src/types.ts [hunk 1]
```diff
@@ -47,5 +47,6 @@ export type NativeCurrency = {
 
 export type RpcUrls = {
   http: readonly string[];
+  headers?: object;
```
Comment (fracek) [inline packages/chains/src/types.ts:50]: Not convinced by this because headers should be dependant on the specific provider used.

**Comment by fracek** [comment]
I think this PR shows a flaw in the current design. IMHO I should change `StarknetConfig.providers` to be either a mapping between chainId and a `JsonRpcProvider` or a function that takes a chain and returns the `JsonRpcProvider`.

If you're ok with that maybe you can update the PR to reflect that design?

**Comment by fracek** [comment]
Actually, forget option 2. The `providers` argument should be renamed `provider` and accepts a `ChainProviderFactory`. If users want chain specific providers, we can provide a higher-level provider to combine two providers.

**Comment by IvanLudvig** [comment]
I agree that the current design is flawed and I implemented your idea. Can you take a look, please? @fracek

**Comment by fracek** [comment]
Looks good! Much cleaner now. Please fix the tests and I'll merge it.

**Comment by IvanLudvig** [comment]
I fixed the tests

**Comment by fracek** [comment]
Perfect, thank you! Are you on onlydust? How much time did you spend on this PR?

**Comment by IvanLudvig** [comment]
I just joined onlydust. I spent 3 hours

**Pull request merged**